docs/proposals/vip-4.md

```diff
@@ -0,0 +1,248 @@
+
+* The `SERVING` step is also called "version swap". If there are Da Vinci Clients, they will ingest and swap on their
+  own, and the child controller of that region will wait until all DVC instances have swapped (started serving) before
+  enacting the region-level swap, after which the servers will also start serving the new store-version to clients which
+  perform remote queries. That is why the "DVC Read Traffic SERVING" step is a dependency for the "Server Read Traffic
+  SERVING" step.
+
+## Problem Statement 
+
+Pushing to all regions in parallel makes the push faster, but it also means that if the push causes an issue, the impact
+is going to be global (affecting all regions). It would be desirable to have certain checks and balances that reduce the
+blast radius in cases where the content of the push causes issues. Examples of such issues include:
+
+* Data quality issues:
+  * Incomplete data (due to issues in data generation logic, or in upstream datasets).
+  * Change in semantics (e.g. some embeddings trained by a new ML model / weights / params / etc. are incompatible with 
+    that used at inference-time).
+* Schema-related issues:
+  * Some optional fields which used to always be populated get deleted (or get populated with null) and the reading app 
+    fails due to lack of null checking. This kind of app-side bug can happen even though a schema evolution is fully 
+    compatible.
+* Infra issues:
+  * Larger payloads take more resources, resulting in lack of capacity and thus latency degradation.
+  * Certain types of yet-unknown infra bugs are somehow triggered by a data push.
+
+Tighter control of how the new version of the dataset is deployed to each region could allow us to catch issues while
+only one region is affected, and abort deploying to other regions. See Scope, below, for specific examples of flow 
+control strategies.
+
+In addition, we would like to make it easier to integrate the push job into proprietary monitoring systems such that 
+each region getting data deployed to it results in events getting emitted or other observability actions.
+
+## Scope
+
+This proposal is about full push jobs. Incremental pushes and nearline writes are out of scope. At the time of 
+submitting this proposal, it is undetermined whether this work will apply to stream reprocessing jobs. In terms of 
+priority, we care mostly about supporting full pushes from offline grids, and it may be fine to leave stream 
+reprocessing out of scope, although depending on the design details we choose, we may be able to support stream 
+reprocessing "for free" as well.
+
+The goal is for lifecycle hooks to achieve the following use cases:
+
+* Orchestrate how data is served in each region, including:
+  * Ensuring a minimum delay (e.g. 1 hour) between each region beginning to serve the new store-version.
+  * Delaying the swapping of a new store-version to be within some time of day (e.g. during "business hours").
```

Comment:
Do we treat business hour range as a global setting or per-store setting? Different customers might have different assumption about the range of the business hours.

Reply:
That is not specified as part of this proposal, but the proposal supports full flexibility, including:

1. Mandatory hooks.
2. Opt-in (disabled by default, can be enabled via the per-store hook config bag).
3. Opt-out (enabled by default, can be disabled via the per-store hook config bag).